R/cal-validate.R

```diff
@@ -0,0 +1,239 @@
+                              location_params = 1,
+                              estimate = dplyr::starts_with(".pred_"),
+                              parameters = NULL,
+                              metrics = NULL,
+                              summarize = TRUE,
+                              ...) {
+  UseMethod("cal_validate_beta")
+}
+
+#' @export
+cal_validate_beta.rset <- function(.data,
+                                   truth = NULL,
+                                   shape_params = 2,
+                                   location_params = 1,
+                                   estimate = dplyr::starts_with(".pred_"),
+                                   parameters = NULL,
+                                   metrics = NULL,
+                                   summarize = TRUE,
+                                   ...) {
+  cal_validate(
+    rset = .data,
+    truth = {{ truth }},
+    estimate = {{ estimate }},
+    cal_function = "beta",
+    metrics = metrics,
+    summarize = summarize,
+    shape_params = shape_params,
+    location_params = location_params,
+    ...
+  )
+}
+
+cal_validate <- function(rset,
+                         truth = NULL,
+                         estimate = NULL,
+                         cal_function = NULL,
+                         metrics = NULL,
+                         summarize = TRUE,
+                         ...) {
+  if (is.null(cal_function)) rlang::abort("No calibration function provided")
+
+  if (is.null(metrics)) {
+    metrics <- yardstick::metric_set(
+      yardstick::brier_class
+    )
+  }
+
+  data_tr <- purrr::map(rset$splits, rsample::training)
+  data_as <- purrr::map(rset$splits, rsample::assessment)
+
+  if (cal_function == "logistic") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_logistic,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  if (cal_function == "isotonic") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_isotonic,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  if (cal_function == "beta") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_beta,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  estimate_col <- cals[[1]]$levels[[1]]
+
+  applied <- seq_along(data_as) %>%
+    map(
+      ~ {
+        val <- cal_apply(data_as[[.x]], cals[[.x]])
+        stats_after <- metrics(val, truth = {{ truth }}, estimate_col)
+        stats_before <- metrics(data_as[[.x]], truth = {{ truth }}, estimate_col)
+
+        list(
+          val = val,
+          stats_after = stats_after,
+          stats_before = stats_before
+        )
+      }
+    ) %>%
+    purrr::transpose()
+
+  ret <- dplyr::mutate(
+    rset,
+    calibration = cals,
+    validation = applied$val,
+    stats_after = applied$stats_after,
+    stats_before = applied$stats_before
+  )
+
+  if (summarize) ret <- summarize_validation(ret)
```

Comment:
I think that we should return the results in a tidy format like this: 

```
# A tibble: 4 × 4
  .metric     .estimator stage   .estimate
  <chr>       <chr>      <chr>       <dbl>
1 brier_class binary     before     0.140
2 brier_class binary     after      0.142
3 roc_auc     binary     before     0.873
4 roc_auc     binary     after      0.869
```

when `summarize = TRUE`.

For `summarize = FALSE`, return that structure in the `.metrics` slot. Also, I think that we should not automatically return the  `calibration` and `validation` columns. Maybe have a `save_details` option that will do this. 

The unsummarized structure looked a lot like a `tune_results` object (but is not). Just thinking that we might want a separate class to add on? What do you think?

Reply:
Ok, looking at that tibble ☝️ , maybe use "uncalibrated" instead of "before" and "calibrated" instead of "after".

Reply:
> For summarize = FALSE, return that structure in the .metrics slot
Does that mean we output an S4 object? as in `slot()`?

Reply:
No, I meant "slot" in the general sense of "put it in the list somewhere" 😄

Reply:
Ok, I went ahead and added an exported function that will calculate and display de summarization, that way we don't have to figure out where to "save it" as part of the tibble

```
library(yardstick)
#> For binary classification, the first factor level is assumed to be the event.
#> Use the argument `event_level = "second"` to alter this as needed.
library(magrittr)
suppressPackageStartupMessages(library(probably))

x <- segment_logistic %>%
  rsample::vfold_cv() %>%
  cal_validate_logistic(Class, summarize = FALSE, metrics = metric_set(brier_class, roc_auc))

x
#> #  10-fold cross-validation 
#> # A tibble: 10 × 4
#>    splits            id     stats_after      stats_before    
#>    <list>            <chr>  <list>           <list>          
#>  1 <split [909/101]> Fold01 <tibble [2 × 4]> <tibble [2 × 4]>
#>  2 <split [909/101]> Fold02 <tibble [2 × 4]> <tibble [2 × 4]>
#>  3 <split [909/101]> Fold03 <tibble [2 × 4]> <tibble [2 × 4]>
#>  4 <split [909/101]> Fold04 <tibble [2 × 4]> <tibble [2 × 4]>
#>  5 <split [909/101]> Fold05 <tibble [2 × 4]> <tibble [2 × 4]>
#>  6 <split [909/101]> Fold06 <tibble [2 × 4]> <tibble [2 × 4]>
#>  7 <split [909/101]> Fold07 <tibble [2 × 4]> <tibble [2 × 4]>
#>  8 <split [909/101]> Fold08 <tibble [2 × 4]> <tibble [2 × 4]>
#>  9 <split [909/101]> Fold09 <tibble [2 × 4]> <tibble [2 × 4]>
#> 10 <split [909/101]> Fold10 <tibble [2 × 4]> <tibble [2 × 4]>

cal_validate_summarize(x)
#> # A tibble: 4 × 5
#>   .metric     .estimator direction stage        .estimate
#>   <chr>       <chr>      <chr>     <chr>            <dbl>
#> 1 brier_class binary     minimize  uncalibrated    0.140 
#> 2 brier_class binary     minimize  calibrated      0.0679
#> 3 roc_auc     binary     maximize  uncalibrated    0.873 
#> 4 roc_auc     binary     maximize  calibrated      1
```